File: examples/scf_grad.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include "cint.h"

#include "f2c.h"
#include "blaswrap.h"

#define N_STO 3

int __enzyme_autodiff(void *, ...);

int cint1e_ovlp_cart(double *buf, int *shls,
                    int *atm, int natm, int *bas, int nbas, double *env);

int cint1e_kin_cart(double *buf, int *shls,
                    int *atm, int natm, int *bas, int nbas, double *env);

int cint1e_nuc_cart(double *buf, int *shls,
                    int *atm, int natm, int *bas, int nbas, double *env);

extern int dsyev_(char *jobz, char *uplo, integer *n, doublereal *a, 
    integer *lda, doublereal *w, doublereal *work, integer *lwork, 
	integer *info);

extern int enzyme_dup;
extern int enzyme_out;
extern int enzyme_const;

typedef struct array {
    double * m;
    int row;
    int col;
} Array;

typedef struct Array_two {
    double * m;
    int x;
    int y;
    int w;
    int z;
} Array_two;

void free_arr(Array * arr) {
    free(arr->m);
    free(arr);
}

void free_arr_two(Array_two * arr) {
    free(arr->m);
    free(arr);
}

void read_arrays(FILE * file, int natm, int nbas, int ** atm, int ** bas, double ** env)
{
	double num;
	int number;
	int count = 0;

	while (count < natm * ATM_SLOTS) {
        fscanf(file, "%d", &number);
		(*atm)[count] = number;
		count++;
	}

	count = 0;
	while (count < nbas * BAS_SLOTS) {
        fscanf(file, "%d", &number);
		(*bas)[count] = number;
		count++;
	}

	count = 0;
	while (fscanf(file, "%lf", &num) != EOF) {
		(*env)[count] = num;
		count++;
	}
	fclose(file);
}

Array * c_arr(int x, int y) 
{   
    Array * arr = malloc(sizeof(Array));

    double * mat = malloc(sizeof(double) * (x * y));
    memset(mat, 0, sizeof(double) * (x * y));

    arr->m = mat;
    arr->row = x;
    arr->col = y;

    return arr;
}

Array_two * c_arr_doub(int x, int y, int w, int z) 
{   
    Array_two * arr = malloc(sizeof(Array_two));

    double * mat = malloc(sizeof(double) * (x * y * w * z));
    memset(mat, 0, sizeof(double) * (x * y * w * z));

    arr->m = mat;
    arr->x = x;
    arr->y = y;
    arr->w = w;
    arr->z = z;

    return arr;
}

void dcopy(Array * dest, Array source) {
    for (int i = 0; i < source.col * source.row; i++) {
        dest->m[i] = source.m[i];
    }
}

void matmult(int n, double * A, double * B, double * C) {
    // cij = aik bkj
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            for (int k = 0; k < n; k++) {
                C[i * n + j] += A[i * n + k] * B[k * n + j];
            }
        }
    }
}

void transpose(int n, double * C, double * Ct) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            Ct[i * n + j] = C[j * n + i];
        }
    }
}

void integrals(int natm, int nbas, int * atm, int * bas, double * env, Array ** S, Array ** T, Array ** V, Array ** H, Array_two ** two) 
{   
    *S = c_arr(nbas, nbas);
    *T = c_arr(nbas, nbas);
    *V = c_arr(nbas, nbas);
    *H = c_arr(nbas, nbas);
    *two = c_arr_doub(nbas, nbas, nbas, nbas);

    int di, dj, dk, dl;
    int shls[4];
    
    double * buf;
    for (int i = 0; i < nbas; i++) {
        for (int j = 0; j < nbas; j++) {
            shls[0] = i; di = CINTcgto_cart(i, bas);
            shls[1] = j; dj = CINTcgto_cart(j, bas);

            buf = malloc(sizeof(double) * di * dj);
            cint1e_ovlp_cart(buf, shls, atm, natm, bas, nbas, env);
            (*S)->m[i * (*S)->row + j] = buf[0];

            cint1e_kin_cart(buf, shls, atm, natm, bas, nbas, env);
            (*T)->m[i * (*T)->row + j] = buf[0];

            cint1e_nuc_cart(buf, shls, atm, natm, bas, nbas, env);
            (*V)->m[i * (*V)->row + j] = buf[0];
            free(buf);

            (*H)->m[i * (*T)->row + j] = (*T)->m[i * (*T)->row + j] + (*V)->m[i * (*V)->row + j];

            for (int k = 0; k < nbas; k++) {
                for (int l = 0; l < nbas; l++) {
                    shls[2] = k; dk = CINTcgto_cart(k, bas);
                    shls[3] = l; dl = CINTcgto_cart(l, bas);

                    buf = malloc(sizeof(double) * di * dj * dk * dl);
                    cint2e_cart(buf, shls, atm, natm, bas, nbas, env, NULL);
                    (*two)->m[(int) (i*pow(nbas, 3) + j*pow(nbas, 2) + k*nbas + l)] = buf[0];
                    free(buf);
                }
            }
        }
    }
}

void find_X(Array S, Array ** X, Array ** X_dag) 
{
    Array * eig = c_arr(1, S.col);
    Array * U = c_arr(S.row, S.col);
    dcopy(U, S);

    char jobz = 'V', uplo = 'U';
    integer lda = 2, n = 2, info = 8, lwork = 6;
    double w[2], work[6];

    dsyev_(&jobz, &uplo, &n, U->m, &lda, w, work, &lwork, &info);

    if(info > 0) {
        printf("The algorithm failed to compute eigenvalues.\n");
        exit(1);
    }

    eig->m = w;

    Array * diag_eig = c_arr(S.col, S.col);
    for (int i = 0; i < S.col; i++) {
        for (int j = 0; j < S.col; j++) {
            if (i == j) {
                diag_eig->m[2*i + j] = pow(eig->m[i], -0.5);
            }
        }
    }

    *X = c_arr(U->row, diag_eig->row);
    matmult(U->row, U->m, diag_eig->m, (*X)->m);

    *X_dag = c_arr((*X)->row, (*X)->col);
    transpose((*X)->row, (*X)->m, (*X_dag)->m);
}

void calc_F(int nbas, Array P, Array_two two, Array H, Array ** G, Array ** F) 
{
    *G = c_arr(nbas, nbas);
    *F = c_arr(nbas, nbas);

    for (int mu = 0; mu < nbas; mu++) {
        for (int nu = 0; nu < nbas; nu++) {
            for (int la = 0; la < nbas; la++) {
                for (int sig = 0; sig < nbas; sig++) {
                    (*G)->m[mu * (*G)->row + nu] += P.m[la * P.row + sig] 
                        * (two.m[(int) (mu*pow(nbas, 3) + nu*pow(nbas, 2) + sig*nbas + la)] 
                        - 0.5 * two.m[(int) (mu*pow(nbas, 3) + la*pow(nbas, 2) + sig*nbas + nu)]);
                }
            }

            (*F)->m[mu * (*F)->row + nu] += (*G)->m[mu * (*G)->row + nu] + H.m[mu * H.row + nu];
        }
    }
}

void calc_Fprime(Array F, Array X, Array X_dag, Array ** Fprime) 
{
    Array * inter = c_arr(X_dag.row, F.col);
    matmult(X_dag.row, X_dag.m, F.m, inter->m);

    *Fprime = c_arr(inter->row, X.col);
    matmult(inter->row, inter->m, X.m, (*Fprime)->m);

    free_arr(inter);
}

void diag_F(Array Fprime, Array X, Array ** C, Array ** epsilon)
{
    Array * U = c_arr(Fprime.row, Fprime.col);
    dcopy(U, Fprime);

    char jobz = 'V', uplo = 'U';
    integer lda = 2, n = 2, info = 8, lwork = 6;
    double w[2], work[6];

    dsyev_(&jobz, &uplo, &n, U->m, &lda, w, work, &lwork, &info);

    if(info > 0) {
        printf("The algorithm failed to compute eigenvalues.\n");
        exit(1);
    }

    Array * Udag = c_arr(U->row, U->col);
    transpose(Udag->row, U->m, Udag->m);
    
    Array * inter = c_arr(Udag->row, Fprime.col);
    matmult(Udag->row, Udag->m, Fprime.m, inter->m);
    
    Array * f = c_arr(inter->row, U->col);
    matmult(inter->row, inter->m, U->m, f->m);

    Array * Cprime = c_arr(U->row, U->col);
    dcopy(Cprime, *U);

    *epsilon = c_arr(f->row, f->col);
    dcopy(*epsilon, *f);

    *C = c_arr(X.row, Cprime->col);
    matmult(X.row, X.m, Cprime->m, (*C)->m);
    
    free_arr(U);
    free_arr(Udag);
    free_arr(inter);
    free_arr(f);
    free_arr(Cprime);
}

Array * calc_P(int nbas, int nelec, Array C) 
{
    Array * P = c_arr(nbas, nbas);
    for (int mu = 0; mu < nbas; mu++) {
        for (int nu = 0; nu < nbas; nu++) {
            for (int i = 0; i < (nelec / 2); i++) {
                // printf("%lf\n", 2.0 * C.m[mu * C.row + i] * C.m[nu * C.row + i]);
                P->m[mu * P->row + nu] += 2.0 * C.m[mu * C.row + i] * C.m[nu * C.row + i];
            }
        }
    }

    return P;
}

double f_delta(int nbas, Array P, Array P_old) 
{
    double delta = 0;
    for (int mu = 0; mu < nbas; mu++) {
        for (int nu = 0; nu < nbas; nu++) {
            delta += pow(P.m[mu * P.row + nu] - P_old.m[mu * P.row + nu], 2.0);
        }
    }
    delta = pow(delta, 0.5) / 2.0;

    return delta;
}

double norm(int * atm, double * env, int i, int j) 
{
    double xi = env[atm[i*6 + 1]];
    double xj = env[atm[j*6 + 1]];

    double yi = env[atm[i*6 + 1] + 1];
    double yj = env[atm[j*6 + 1] + 1];

    double zi = env[atm[i*6 + 1] + 2];
    double zj = env[atm[j*6 + 1] + 2];

    return pow(pow(xi - xj, 2) + pow(yi - yj, 2) + pow(zi - zj, 2), 0.5); 
}

Array * RHF(int natm, int nbas, int nelec, int * atm, int * bas, double * env, int imax, double conv)
{   
    Array * S, * T, * V, * H;
    Array_two *two;
    integrals(natm, nbas, atm, bas, env, &S, &T, &V, &H, &two);

    Array * X, * X_dag;
    find_X(*S, &X, &X_dag);

    Array * P = c_arr(nbas, nbas);
    for (int i = 0; i < nbas; i++) {
        for (int j = 0; j < nbas; j++) {
            if (i == j) {
                P->m[i * P->row + j] = 1.0;
            }
        }
    }

    int i = 0;
    double delta = 1.0;

    Array * Pold;
    Array * G, * F;
    Array * Fprime;
    Array * C, * epsilon;
    
    while (delta > conv && i < imax) {
        Pold = P;
        
        calc_F(nbas, *P, *two, *H, &G, &F);
        calc_Fprime(*F, *X, *X_dag, &Fprime);
        diag_F(*Fprime, *X, &C, &epsilon);

        P = calc_P(nbas, nelec, *C);

        delta = f_delta(nbas, *P, *Pold);
        i++;
    }

    if (delta > conv && i == imax) {
        printf("did not converge\n");
        exit(1);
    }

    return P;
}

void energy(double * E, int natm, int nbas, int * atm, int * bas, double * env, Array * P)
{
    Array * S, * T, * V, * H;
    Array_two *two;
    integrals(natm, nbas, atm, bas, env, &S, &T, &V, &H, &two);

    Array * G, * F;    
    calc_F(nbas, *P, *two, *H, &G, &F);

    double E0 = 0.0;
    for (int mu = 0; mu < nbas; mu++) {
        for (int nu = 0; nu < nbas; nu++) {
            E0 += 0.5 * P->m[mu * P->row + nu] * (H->m[mu * H->row + nu] + F->m[mu * F->row + nu]);
        }
    }

    double Enuc = 0.0;
    for (int i = 0; i < natm; i++) {
        for (int j = 0; j < natm; j++) {
            if (i > j) {
                Enuc += ((double) (atm[i*6 + 0] * atm[j*6 + 0]))/(norm(atm, env, i, j));
            }
        }
    }

    *E = Enuc + E0;
}

void energy_diff(double * E, double * dE, int natm, int nbas, int * atm, int * bas, double * env, double * denv, Array * P) {
	__enzyme_autodiff((void *) energy,
        enzyme_dup, E, dE,	
        enzyme_const, natm,
        enzyme_const, nbas,
        enzyme_const, atm,
        enzyme_const, bas, 
        enzyme_dup, env, denv,
        enzyme_const, P);
}

int main()
{
    int natm = 2;
    int nbas = 2;
    
    int nelec = 2; // ?
    int nshells = 1; // ?

    int * atm = malloc(sizeof(int) * natm * ATM_SLOTS);
    int * bas = malloc(sizeof(int) * nbas * BAS_SLOTS);
    double * env = malloc(sizeof(double) * 10000);

    FILE * file = fopen("/u/jpmedina/libcint/molecules/h2/basis.txt", "r");
    read_arrays(file, natm, nbas, &atm, &bas, &env);

    // RHF
    int imax = 20;
    double conv = 0.000001;

    Array * P = RHF(natm, nbas, nelec, atm, bas, env, imax, conv);

    double E;
    double dE = 1.0;
    double * denv = malloc(sizeof(double) * 10000);

    energy_diff(&E, &dE, natm, nbas, atm, bas, env, denv, P);
    printf("E: %lf\n", E);
    printf("denv:\n");
    for (int k = 28; k < 34; k++) {
        printf("%f ", denv[k]);
    }
    printf("\n");

    printf("finite diff:\n");
    
    double grad;
    double E1, E2;
    double h = 0.000001;
    
    for (int k = 28; k < 34; k++) {
        env[k] += h;
        energy(&E1, natm, nbas, atm, bas, env, P);
        env[k] -= 2.0*h;
        energy(&E2, natm, nbas, atm, bas, env, P);
        env[k] += h;

        grad = (E1 - E2)/(2.0*h);

        printf("%lf ", grad);
    }
    printf("\n");

    return 0;
}

File: librint/grad/h2/scf_grad.rs
#![feature(autodiff)]

use librint::cint_bas::CINTcgto_cart;
use librint::cint1e::cint1e_ovlp_cart;
use librint::cint1e::cint1e_nuc_cart;
use librint::intor1::cint1e_kin_cart;
use librint::cint2e::cint2e_cart;

use faer::mat;
use faer::linalg::solvers::Eigendecomposition;
use faer::complex_native::c64;

pub const ATM_SLOTS: usize = 6;
pub const BAS_SLOTS: usize = 8;

pub const N_STO: i32 = 3;

fn matmult(
    n: usize,
    A: &mut [f64],
    B: &mut [f64],
    C: &mut [f64],
) {
    // cij = aik bkj
    C.fill(0.0);
    for i in 0..n {
        for j in 0..n {
            for k in 0..n {
                C[i * n + j] += A[i * n + k] * B[k * n + j];
            }
        }
    }
}

fn transpose(
    n: usize,
    C: &mut [f64],
    Ct: &mut [f64],
) {
    for i in 0..n {
        for j in 0..n {
            Ct[i * n + j] = C[j * n + i]
        }
    }
}

fn dcopy(
    n: usize,
    A: &mut [f64],
    Ac: &mut [f64],
) {
    for i in 0..n {
        for j in 0..n {
            Ac[i * n + j] = A[i * n + j];
        }
    }
}

fn zero(

) {

}

fn print_mat(
    n: usize,
    C: &mut [f64],
) {
    for i in 0..n {
        for j in 0..n {
            print!("{} ", C[i * n + j]);
        }
        println!();
    }
}

fn integrals(
    natm: usize,
    nbas: usize,
    atm: &mut [i32],
    bas: &mut [i32],
    env: &mut [f64],
    S: &mut [f64],
    T: &mut [f64],
    V: &mut [f64],
    H: &mut [f64],
    two: &mut [f64],
) {
    let mut buf;

    let mut shls: [i32; 4] = [0, 0, 0, 0];

    for i in 0..nbas {
        for j in 0..nbas {
            shls[0] = i as i32;
            shls[1] = j as i32;
            
            let di = CINTcgto_cart(i as usize, &bas);
            let dj = CINTcgto_cart(j as usize, &bas);

            buf = vec![0.0; (di * dj) as usize];
            cint1e_ovlp_cart(&mut buf, &mut shls, atm, natm as i32, bas, nbas as i32, env);
            S[i*nbas + j] = buf[0];

            cint1e_kin_cart(&mut buf, &mut shls, atm, natm as i32, bas, nbas as i32, env);
            T[i*nbas + j] = buf[0];

            cint1e_nuc_cart(&mut buf, &mut shls, atm, natm as i32, bas, nbas as i32, env);
            V[i*nbas + j] = buf[0];

            H[i*nbas + j] = T[i*nbas + j] + V[i*nbas + j];

            for k in 0..nbas {
                for l in 0..nbas {
                    shls[2] = k as i32;
                    shls[3] = l as i32;
        
                    let dk = CINTcgto_cart(k as usize, &bas);
                    let dl = CINTcgto_cart(l as usize, &bas);

                    buf = vec![0.0; (di * dj * dk * dl) as usize];
                    cint2e_cart(&mut buf, &mut shls, atm, natm as i32, bas, nbas as i32, env);
                    two[i*nbas.pow(3) + j*nbas.pow(2) + k*nbas.pow(1) + l] = buf[0];
                }
            }
        }
    }
}

fn find_X(
    nbas: usize,
    S: &mut [f64],
    X: &mut [f64],
    Xdag: &mut [f64],
) {
    let s_mat = mat::from_column_major_slice::<f64>(&S, nbas, nbas);
    let eig_decomp: Eigendecomposition<c64> = Eigendecomposition::new_from_real(s_mat);
    let eigenvalues = eig_decomp.s();
    let eigenvectors = eig_decomp.u();

    let mut U = vec![0.0; nbas * nbas];

    for i in 0..eigenvectors.nrows() {
        for j in 0..eigenvectors.ncols() {
            U[i * nbas + j] = eigenvectors.read(i, j).re();
        }
    }

    let eign = eigenvalues.column_vector();
    let mut lamb = vec![0.0; nbas * nbas];
    for i in 0..nbas {
        for j in 0..nbas {
            if i == j {
                let li: f64 = eign.read(i).re();
                lamb[i * nbas + j] = li.powf(-0.5);
            }
        }
    }

    matmult(nbas, &mut U, &mut lamb, X);
    transpose(nbas, X, Xdag);
}

fn calc_F(
    nbas: usize,
    P: &mut [f64],
    two: &mut [f64],
    H: &mut [f64],
    G: &mut [f64],
    F: &mut [f64],
) {
    G.fill(0.0);
    F.fill(0.0);

    for mu in 0..nbas {
        for nu in 0..nbas {
            for la in 0..nbas {
                for sig in 0..nbas {
                    G[mu*nbas + nu] += P[la*nbas + sig]
                        * (two[mu*nbas.pow(3) + nu*nbas.pow(2) + sig*nbas + la]
                        - 0.5 * two[mu*nbas.pow(3) + la*nbas.pow(2) + sig*nbas + nu]);
                }
            }

            F[mu*nbas + nu] += G[mu*nbas + nu] + H[mu*nbas + nu];
        }
    }
}

fn calc_Fprime(
    nbas: usize,
    F: &mut [f64],
    X: &mut [f64],
    Xdag: &mut [f64],
    mut Fprime: &mut [f64],
) {
    let mut inter = vec![0.0; nbas * nbas];
    matmult(nbas, Xdag, F, &mut inter);
    matmult(nbas, &mut inter, X, &mut Fprime);
}

fn diag_F(
    nbas: usize,
    Fprime: &mut [f64],
    X: &mut [f64],
    C: &mut [f64],
    epsilon: &mut [f64],
) {
    // diagonalize Fprime
    let fprime_mat = mat::from_column_major_slice::<f64>(&Fprime, nbas, nbas);
    let eig_decomp: Eigendecomposition<c64> = Eigendecomposition::new_from_real(fprime_mat);
    let eigenvalues = eig_decomp.s();
    let eigenvectors = eig_decomp.u();

    let mut U = vec![0.0; nbas * nbas];
    for i in 0..eigenvectors.nrows() {
        for j in 0..eigenvectors.ncols() {
            U[i * nbas + j] = eigenvectors.read(i, j).re();
        }
    }

    let mut Udag = vec![0.0; nbas * nbas];
    transpose(nbas, &mut U, &mut Udag);

    let mut inter = vec![0.0; nbas * nbas];
    matmult(nbas, &mut Udag, Fprime, &mut inter); 

    let mut f = vec![0.0; nbas * nbas];
    matmult(nbas, &mut inter, &mut U, &mut f);

    let mut Cprime = vec![0.0; nbas * nbas];
    dcopy(nbas, &mut U, &mut Cprime);

    dcopy(nbas, &mut f, epsilon);

    matmult(nbas, X, &mut Cprime, C);
}

fn calc_P(
    nbas: usize,
    nelec: usize,
    C: &mut [f64],
) -> Vec<f64> {
    let mut P = vec![0.0; nbas * nbas];
    for mu in 0..nbas {
        for nu in 0..nbas {
            for i in 0..(nelec/2) {
                P[mu * nbas + nu] += 2.0 * C[mu * nbas + i] * C[nu * nbas + i];
            }
        }
    }
    return P;
}

fn f_delta(
    nbas: usize,
    P: &mut [f64],
    Pold: &mut [f64],
) -> f64 {
    let mut delta: f64 = 0.0;
    for mu in 0..nbas {
        for nu in 0..nbas {
            delta += (P[mu * nbas + nu] - Pold[mu * nbas + nu]).powf(2.0);
        }
    }
    delta = delta.powf(0.5) / 2.0;
    return delta;
}

fn norm(
    atm: &mut [i32],
    env: &mut [f64],
    i: usize,
    j: usize,
) -> f64 {
    let xi: f64 = env[(atm[i*6 + 1]) as usize];
    let xj: f64 = env[(atm[j*6 + 1]) as usize];

    let yi: f64 = env[(atm[i*6 + 1] + 1) as usize];
    let yj: f64 = env[(atm[j*6 + 1] + 1) as usize];

    let zi: f64 = env[(atm[i*6 + 1] + 2) as usize];
    let zj: f64 = env[(atm[j*6 + 1] + 2) as usize];

    return ((xi - xj).powf(2.0) + (yi - yj).powf(2.0) + (zi - zj).powf(2.0)).powf(0.5);
}

fn RHF(
    natm: usize,
    nbas: usize,
    nelec: usize,
    atm: &mut [i32],
    bas: &mut [i32],
    env: &mut [f64],
    imax: i32,
    conv: f64,
) -> Vec<f64> {
    let mut S = vec![0.0; nbas * nbas];
    let mut T = vec![0.0; nbas * nbas];
    let mut V = vec![0.0; nbas * nbas];
    let mut H = vec![0.0; nbas * nbas];
    let mut two = vec![0.0; nbas * nbas * nbas * nbas];
    integrals(natm, nbas, atm, bas, env, &mut S, &mut T, &mut V, &mut H, &mut two);

    let mut X = vec![0.0; nbas * nbas];
    let mut Xdag = vec![0.0; nbas * nbas];
    find_X(nbas, &mut S, &mut X, &mut Xdag);

    let mut P = vec![0.0; nbas * nbas];
    for i in 0..nbas {
        for j in 0..nbas {
            if i == j {
                P[i*nbas + j] = 1.0;
            }
        }
    }

    let mut i: i32 = 0;
    let mut delta: f64 = 1.0;

    let mut Pold = vec![0.0; nbas * nbas];
    let mut G = vec![0.0; nbas * nbas];
    let mut F = vec![0.0; nbas * nbas];
    let mut Fprime = vec![0.0; nbas * nbas];
    let mut C = vec![0.0; nbas * nbas];
    let mut epsilon = vec![0.0; nbas * nbas];

    while delta > conv && i < imax {
        dcopy(nbas, &mut P, &mut Pold);

        calc_F(nbas, &mut P, &mut two, &mut H, &mut G, &mut F);
        calc_Fprime(nbas, &mut F, &mut X, &mut Xdag, &mut Fprime);
        diag_F(nbas, &mut Fprime, &mut X, &mut C, &mut epsilon);

        P = calc_P(nbas, nelec, &mut C);
        delta = f_delta(nbas, &mut P, &mut Pold);
        i += 1;
    }

    println!("Conv P:");
    for i in 0..nbas {
        for j in 0..nbas {
            print!("{} ", P[i*nbas + j]);
        }
        println!();
    }

    if delta > conv && i == imax {
        println!("did not converge");
    }

    return P;
}

#[autodiff(denergy, Reverse, Duplicated, Const, Const, Const, Const, Duplicated, Const)]
fn energy(
    E: &mut f64,
    natm: usize,
    nbas: usize,
    atm: &mut [i32],
    bas: &mut [i32],
    env: &mut [f64],
    P: &mut [f64],
) {
    let mut S = vec![0.0; nbas * nbas];
    let mut T = vec![0.0; nbas * nbas];
    let mut V = vec![0.0; nbas * nbas];
    let mut H = vec![0.0; nbas * nbas];
    let mut two = vec![0.0; nbas * nbas * nbas * nbas];
    integrals(natm, nbas, atm, bas, env, &mut S, &mut T, &mut V, &mut H, &mut two);

    let mut G = vec![0.0; nbas * nbas];
    let mut F = vec![0.0; nbas * nbas];
    calc_F(nbas, P, &mut two, &mut H, &mut G, &mut F);

    let mut E0: f64 = 0.0;
    for mu in 0..nbas {
        for nu in 0..nbas {
            E0 += 0.5 * P[mu*nbas + nu] * (H[mu*nbas + nu]  + F[mu*nbas + nu]);
        }
    }

    let mut Enuc: f64 = 0.0;
    for i in 0..nbas {
        for j in 0..nbas {
            if i > j {
                Enuc += (atm[i*6 + 0] * atm[j*6 + 0]) as f64 / (norm(atm, env, i, j));
            }
        }
    }

    *E = Enuc + E0;
}

fn main() {
    const natm: usize = 2;
    const nbas: usize = 2;

    const nelec: usize = 2;

    let mut atm: [i32; natm * ATM_SLOTS] = [1, 20, 1, 23, 0, 0, 1, 24, 1, 27, 0, 0];
    let mut bas: [i32; nbas * BAS_SLOTS] = [0, 0, 3, 1, 0, 28, 31, 0, 1, 0, 3, 1, 0, 28, 31, 0];
    let mut env: [f64; 34] = [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 
        -1.5117809, 0., 0., 0., 1.5117809, 0., 3.42525091, 0.62391373, 0.1688554, 0.98170675, 0.94946401, 0.29590645];
    let mut denv: [f64; 34] = [0.0; 34];

    const imax: i32 = 2;
    const conv: f64 = 0.000001;

    let mut P = RHF(natm, nbas, nelec, &mut atm, &mut bas, &mut env, imax, conv);

    let mut Etot: f64 = 0.0;
    let mut dEtot: f64 = 1.0;
    denergy(&mut Etot, &mut dEtot, natm, nbas, &mut atm, &mut bas, &mut env, &mut denv, &mut P);
    println!("Etot: {}", Etot);

    println!("denv:");
    for i in 28..36 {
        print!("{} ", denv[i]);
    }
}